fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeRewriter.java

```diff
@@ -0,0 +1,403 @@
+        Map<ColumnRefOperator, ScalarOperator> dictRefToDictExprMap = Maps.newHashMap();
+        for (Integer stringId : decodeIds.getColumnIds()) {
+            if (!inputIds.contains(stringId)) {
+                continue;
+            }
+            ColumnRefOperator stringRef = factory.getColumnRef(stringId);
+            ColumnRefOperator dictRef = context.stringRefToDictRefMap.get(stringRef);
+            dictRefToStringRefMap.put(dictRef, stringRef);
+        }
+
+        PhysicalDecodeOperator decodeOperator =
+                new PhysicalDecodeOperator(ImmutableMap.copyOf(dictRefToStringRefMap), dictRefToDictExprMap);
+
+        LogicalProperty property = new LogicalProperty(child.getLogicalProperty());
+        ColumnRefSet outputColumns = child.getLogicalProperty().getOutputColumns();
+
+        final ColumnRefSet rewriteOutputColumns = new ColumnRefSet();
+        // rewrite dict column -> string column
+        outputColumns.getStream().map(factory::getColumnRef).map(c -> dictRefToStringRefMap.getOrDefault(c, c))
+                .forEach(rewriteOutputColumns::union);
+        property.setOutputColumns(rewriteOutputColumns);
+
+        // use child's info
+        return OptExpression.builder().with(child).setOp(decodeOperator).setLogicalProperty(property)
+                .setInputs(Lists.newArrayList(child)).build();
+    }
+
+    @Override
+    public OptExpression visit(OptExpression optExpression, ColumnRefSet fragmentUseDictExprs) {
+        PhysicalOperator op = optExpression.getOp().cast();
+        DecodeInfo info = context.operatorDecodeInfo.getOrDefault(op, DecodeInfo.EMPTY);
+        op.setPredicate(rewritePredicate(op.getPredicate(), info.inputStringColumns));
+        op.setProjection(rewriteProjection(op.getProjection(), info.inputStringColumns));
+        return rewriteOptExpression(optExpression, op, info.outputStringColumns);
+    }
+
+    @Override
+    public OptExpression visitPhysicalHashAggregate(OptExpression optExpression, ColumnRefSet fragmentUseDictExprs) {
+        // rewrite multi-stage aggregate
+        PhysicalHashAggregateOperator aggregate = optExpression.getOp().cast();
+        DecodeInfo info = context.operatorDecodeInfo.getOrDefault(aggregate, DecodeInfo.EMPTY);
+        ColumnRefSet inputStringRefs = new ColumnRefSet();
+        inputStringRefs.union(info.inputStringColumns);
+
+        List<ColumnRefOperator> groupBys = aggregate.getGroupBys().stream()
+                .map(c -> inputStringRefs.contains(c) ? context.stringRefToDictRefMap.getOrDefault(c, c) : c)
+                .collect(Collectors.toList());
+        List<ColumnRefOperator> partitions = aggregate.getPartitionByColumns().stream()
+                .map(c -> inputStringRefs.contains(c) ? context.stringRefToDictRefMap.getOrDefault(c, c) : c)
+                .collect(Collectors.toList());
+
+        Map<ColumnRefOperator, CallOperator> aggregations = Maps.newHashMap();
```

Comment:
may have problem ref https://github.com/StarRocks/starrocks/pull/7657

Reply:
really is desgin for bug @packy92 we can refactor aggregationDistinctPos, `distinct` aggregate function depend on `HashMap`'s order in `count(distinct) group by 3-stage mode`, it's really unbelievable

Reply:
amazing and unbelievable...

Reply:
https://github.com/StarRocks/starrocks/issues/37016